### PR TITLE
feat(stm): fixing merkle depth for all circuits

### PR DESCRIPTION
**Pull request by damrobi**
## Content

This PR includes changes to the way the non-recursive circuit checks the merkle path of the elements of the witness. The length of the merkle paths given to the circuit is fixed to `13` to allow for the registration of up to `8192` signers. The merkle path computation for the witness generation pads the paths to the given length of `13` using `0` padding meaning it adds arrays `[0u8; 32]` to the path to fill it. The non-recursive circuit is changed to check of this padding (the `0` values) and ignore them when computing the root of the tree.

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [ ] Crates versions are updated (if relevant)
  - [ ] CHANGELOG file is updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] All check jobs of the CI have succeeded
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [x] No new TODOs introduced

## Comments

Benchmark results (in release mode):
  - Without caching:
    - regular computation (new() function): 1ms
    - depth 5: 1.8ms
    - depth 10: 62ms
    - depth 13: 412ms
  - With caching:
    - regular computation (new() function): 1ms
    - depth 5: 1ms
    - depth 10: 1.5ms
    - depth 13: 2.1ms

## Issue(s)

Relates to #3183

### CODE REVIEW

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 1]
```diff
@@ -42,25 +42,79 @@ impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
...
+    /// Provided a non-empty list of leaves and a depth, generates its corresponding
+    /// `MerkleTree` of this given depth.
+    ///
+    /// Check that the given depth makes sense
+    /// If given depth matches the depth of the given leaves, fallback to new
+    /// If it doesn't match, set the number of nodes to the one needed to create
+    /// the merkle tree with appropriate depth
+    pub(crate) fn new_given_depth(leaves: &[L], merkle_tree_depth: u32) -> MerkleTree<D, L> {
+        let n = leaves.len();
+        assert!(n > 0, "MerkleTree::create() called with no leaves");
```
Comment (hjeljeli32) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: We can have panic here, but maybe this code was not cleaned up yet

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 2]
```diff
@@ -42,25 +42,79 @@ impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
...
+        // TODO: This assert should be replaced with proper error handling once we can return an StmResult
+        assert!(
```
Comment (hjeljeli32) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: same comment as above

**Review by hjeljeli32 — COMMENTED** [review]
Overall the direction is good but maybe the code is still not brushed up, I spotted a few panic locations. Also as Copilot spotted, MERKLE_TREE_DEPTH_FOR_SNARK is set to 8 while some tests/comment assume 13.

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs [hunk 3]
```diff
@@ -38,14 +38,27 @@ pub(crate) fn verify_merkle_path(
...
+        .zip(inputs.merkle_positions.iter().skip(1))
+        .try_fold(first_node, |acc, (x, pos)| {
             let left = std_lib.select(layouter, pos, &acc, x)?;
             let right = std_lib.select(layouter, pos, x, &acc)?;
-            std_lib.poseidon(layouter, &[left, right])
+            let current_node = std_lib.poseidon(layouter, &[left, right])?;
+
+            let is_zero = std_lib.is_zero(layouter, x)?;
+            std_lib.select(layouter, &is_zero, &acc, &current_node)
         })?;
```
[the remark this answers is not in the document]
Comment (damrobi) [reply mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs]: The chances of a 0 value appearing from a legitimate hash is negligible and would probably break the hash function so it can be use as a marker to ignore the siblings.

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 4]
```diff
@@ -225,6 +279,41 @@ impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
...
+            proof.push(h.clone());
+            idx = parent(idx);
+        }
+        println!("base proof length{:?}", proof.len());
+
+        while proof.len() < path_length as usize {
+            proof.push(vec![0u8; 32]);
+        }
+        println!("padded proof length{:?}", proof.len());
+
```
[the remark this answers is not in the document]
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs:244]: This is outdated

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs [hunk 5]
```diff
@@ -38,14 +40,35 @@ pub(crate) fn verify_merkle_path(
...
         ],
     )?;
+    // Compute the first node after the leaves outside the loop
+    // to not ignore a potential padding zero leaf
+    // other zero leaf are ignored
+    let first_sibling = inputs
+        .merkle_siblings
+        .first()
+        .ok_or(anyhow!("Missing siblings"))
+        .map_err(to_synthesis_error)?;
```
Comment (damrobi) [inline mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs:87]: I am not sure how to handle potential errors here. They should never happen as it would mean no sibling exists so a merkle tree of only one node. I think it could be replaced by a safe unwrap but I wanted more opinions.

File: mithril-stm/src/proof_system/halo2_snark/mod.rs [hunk 6]
```diff
@@ -21,6 +21,13 @@ pub(crate) use signer::SnarkProofSigner;
 pub(crate) use single_signature::SingleSignatureForSnark;
 pub(crate) use unsafe_helpers::SnarkSetup;
 
+/// Fixed merkle tree depth used to create the merkle tree
+/// of signers for the SNARK proof
+/// Changing this value means changing the circuit and the circuit verification key
+///
+/// For now set to 13 to allow for 2^13 = 8192 signers
+pub const MERKLE_TREE_DEPTH_FOR_SNARK: u32 = 8;
```
[the remark this answers is not in the document]
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/mod.rs]: Outdated

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 7]
```diff
@@ -42,25 +42,79 @@ impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
...
+        if current_merkle_tree_depth == merkle_tree_depth {
+            return Self::new(leaves);
+        }
+
+        // By using (1 << merkle_tree_depth), we are guaranteed to reach the number of nodes
+        // needed to have a depth of merkle_tree_depth
+        let num_nodes = n + (1 << merkle_tree_depth) - 1;
+
+        Self::create_merkle_tree(leaves, num_nodes)
```
[the remark this answers is not in the document]
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: Outdated

File: mithril-stm/src/proof_system/halo2_snark/proof.rs [hunk 8]
```diff
@@ -414,6 +404,122 @@ mod tests {
...
+        )
+        .unwrap();
+
+        let snark_proof = SnarkProof {
+            circuit_proof,
+            circuit_verification_key: CircuitVerificationKey::new(
+                snark_prover.setup.verification_key.clone(),
+            ),
+            params: clerk.parameters,
+            merkle_tree_depth: 13,
```
[the remark this answers is not in the document]
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/proof.rs]: Outdated

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs (hunk 1 above)
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: This panic was already here. I don't think we want to change it for now but we will need to at some point in the future.

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs (hunk 2 above)
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: This one was removed when I removed some of the early commits

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs (hunk 5 above)
Comment (hjeljeli32) [reply mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs:87]: @damrobi 
I would prefer to have these 2 guards:
- guarantee from STM side during witness construction that we send non-empty Merkle paths
- adding a specific circuit error:
  - MissingMerklePathFirstSibling
  - MissingMerklePathFirstPosition
 
IMO that's safest and can become testable (we can add appropriate golden test for that).

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs [hunk 9]
```diff
@@ -38,14 +40,35 @@ pub(crate) fn verify_merkle_path(
             inputs.lottery_target_value.clone(),
         ],
     )?;
+    // Compute the first node after the leaves outside the loop
+    // to not ignore a potential padding zero leaf
+    // other zero leaf are ignored
+    let first_sibling = inputs
```
Comment (hjeljeli32) [inline mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs:80]: I would prefer to have these 2 guards:
- guarantee from STM side during witness construction that we send non-empty Merkle paths
- adding a specific circuit error:
  - MissingMerklePathFirstSibling
  - MissingMerklePathFirstPosition
 
IMO that's safest and can become testable (we can add appropriate golden test for that).

File: mithril-stm/src/proof_system/halo2_snark/proof.rs [hunk 10]
```diff
@@ -299,12 +309,8 @@ mod tests {
...
 
-    fn create_prover(
-        params: Parameters,
-        number_of_signers: usize,
-        seed: [u8; 32],
-    ) -> SnarkProver<ChaCha20Rng> {
-        let merkle_tree_depth = number_of_signers.next_power_of_two().trailing_zeros();
+    fn create_prover(params: Parameters, seed: [u8; 32]) -> SnarkProver<ChaCha20Rng> {
+        let merkle_tree_depth = 13;
         SnarkProver::try_new_deterministic(&params, merkle_tree_depth, seed).unwrap()
```
[the remark this answers is not in the document]
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/proof.rs]: Comment outdated

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs (hunk 9 above)
Comment (damrobi) [reply mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs:80]: Regarding the circuit part, as we discussed, I used the already existing errors that are returned during the initial verification of the witness in the circuit. Since the merkle path length is already validated at the beginning of the circuit, we should never meet the missing first sibling/position error but it is there just in case.
Regarding the STM check for non-empty path, I updated the path computation to always use a padding that relies on a global constant ([here](https://github.com/input-output-hk/mithril/blob/01372fe88d69af106326f763ccc2f94689bedd52/mithril-stm/src/proof_system/halo2_snark/prover_input.rs#L147-L150)). Do you think this is enough or should I add a specific check after the computation of the path?

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs [hunk 11]
```diff
@@ -20,6 +22,8 @@ pub(crate) struct MerklePathInputs<'a> {
     pub(crate) merkle_siblings: &'a [AssignedNative<CircuitBase>],
     /// Assigned sibling positions for each Merkle path level.
     pub(crate) merkle_positions: &'a [AssignedBit<CircuitBase>],
+    /// Assigned sibling positions for each Merkle path level.
```
Comment (curiecrypt) [inline mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs]:
```suggestion
    /// Expected Merkle tree depth, used for error reporting on empty paths.
```

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 12]
```diff
@@ -225,6 +225,27 @@ impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
 
         MerklePath::new(proof, i)
     }
+
+    #[cfg(feature = "future_snark")]
+    // TODO: remove this allow dead_code directive when function is called or future_snark is activated
+    #[allow(dead_code)]
+    /// Get a path (hashes of siblings of the path to the root node)
```
Comment (curiecrypt) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs:230]:
```suggestion
    #[cfg(feature = "future_snark")]
    /// Get a path (hashes of siblings of the path to the root node)
```

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs
Comment (curiecrypt) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs:222]:
```suggestion
            proof.push(h);
```

Comment (curiecrypt) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs:220]:
```suggestion
                vec![0u8; <D as Digest>::output_size()]
```

I think we need to modify this line like this as Jia's concern.

Comment (curiecrypt) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs]:
```suggestion
            let z = vec![0u8; <D as Digest>::output_size()];
```
similarly

**Review by curiecrypt — APPROVED** [review]
The core design, fixed depth constant + path padding + circuit `is_zero` skip, is sound. There is a blocking item, we need to modify the dynamic depth usages in the rest of the library. The following are the ones I detected:
- `../protocol/aggregate_signature/clerk.rs` line 93
- `../proof_system/halo2_snark/proof.rs` line 234

Other than that LGTM 👍

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs
Comment (jpraynaud) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: I'm not sure that we want to modify the internal implementation of the Merkle tree, as replacing `D::digest([0u8]).to_vec()` with `vec![0u8; <D as Digest>::output_size()]` would create a breaking change in the Merkle root computation.

Comment (jpraynaud) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs:220]: I'm not sure about this or we may introduce a breaking change as Merkle roots will I'm not sure that we want to modify the internal implementation of the Merkle tree, as replacing `D::digest([0u8]).to_vec()` with `vec![0u8; <D as Digest>::output_size()]` would create a breaking change in the Merkle root computation.

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 13]
```diff
@@ -564,6 +585,59 @@ mod tests {
...
+            }
+        }
+
+        #[test]
+        fn padded_path_has_correct_length() {
+            let leaves = make_leaves(4); // natural depth = 2
+            let tree = MerkleTree::<SnarkHash, MerkleTreeSnarkLeaf>::new(&leaves);
+            let path_length = 5u32;
+            let pf = tree.compute_merkle_tree_path_fixed_length(0, path_length);
+            assert_eq!(pf.values.len(), path_length as usize);
```
Comment (jpraynaud) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs:627]:
```suggestion
            let path_length = 5u32;
            
            let pf = tree.compute_merkle_tree_path_fixed_length(0, path_length);
            assert_eq!(pf.values.len(), path_length as usize);
```

File: mithril-stm/src/proof_system/halo2_snark/mod.rs [hunk 14]
```diff
@@ -21,6 +21,13 @@ pub(crate) use signer::SnarkProofSigner;
 pub(crate) use single_signature::SingleSignatureForSnark;
 pub(crate) use unsafe_helpers::SnarkSetup;
 
+/// Fixed merkle tree depth used to create the merkle tree
+/// of signers for the SNARK proof
+/// Changing this value means changing the circuit and the circuit verification key
```
Comment (jpraynaud) [inline mithril-stm/src/proof_system/halo2_snark/mod.rs]:
```suggestion
/// Fixed merkle tree depth used to create the merkle tree
/// of signers for the SNARK proof
///
/// Important: changing this value means changing the circuit and its circuit verification key
```

File: mithril-stm/src/proof_system/halo2_snark/proof.rs [hunk 15]
```diff
@@ -340,6 +335,42 @@ mod tests {
...
+        let path_length = prover_prep.into_witness().first().unwrap().merkle_path.siblings.len();
+
+        assert!(path_length >= current_merkle_tree_depth as usize);
+
+        let snark_proof = prover
+            .aggregate_signatures::<D>(&clerk, &signatures, &message)
+            .unwrap();
+        let result = snark_proof.verify(message.as_slice(), &avk);
+
+        assert!(result.is_ok());
```
Comment (jpraynaud) [inline mithril-stm/src/proof_system/halo2_snark/proof.rs:371]:
```suggestion
        snark_proof.verify(message.as_slice(), &avk).expect("SNARK proof verification should succeed");
```

File: mithril-stm/src/proof_system/halo2_snark/proof.rs [hunk 16]
```diff
@@ -299,13 +302,8 @@ mod tests {
...
 
-    fn create_prover(
-        params: Parameters,
-        number_of_signers: usize,
-        seed: [u8; 32],
-    ) -> SnarkProver<ChaCha20Rng> {
-        let merkle_tree_depth = number_of_signers.next_power_of_two().trailing_zeros();
-        SnarkProver::try_new_deterministic(&params, merkle_tree_depth, seed).unwrap()
+    fn create_prover(params: Parameters, seed: [u8; 32]) -> SnarkProver<ChaCha20Rng> {
+        SnarkProver::try_new_deterministic(&params, MERKLE_TREE_DEPTH_FOR_SNARK, seed).unwrap()
```
Comment (jpraynaud) [inline mithril-stm/src/proof_system/halo2_snark/proof.rs:306]: If this has an impact on test execution time, we can use a smaller depth here.

File: mithril-stm/src/proof_system/halo2_snark/prover_input.rs [hunk 17]
```diff
@@ -141,7 +144,10 @@ impl SnarkProverInput {
                     Ok(None) => continue,
                     Err(_) => continue,
                 };
-                let merkle_path = merkle_tree.compute_merkle_tree_path(sig.signer_index as usize);
+                let merkle_path = merkle_tree.compute_merkle_tree_path_fixed_length(
```
Comment (jpraynaud) [inline mithril-stm/src/proof_system/halo2_snark/prover_input.rs:147]: Modifying this line should probably break a test somewhere 🤔

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs [hunk 18]
```diff
@@ -38,14 +42,41 @@ pub(crate) fn verify_merkle_path(
             inputs.lottery_target_value.clone(),
         ],
     )?;
+    // Compute the first node after the leaves outside the loop
+    // to not ignore a potential padding zero leaf
+    // other zero leaf are ignored
```
Comment (jpraynaud) [inline mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs]: This comment is not very clear. You should create a better description explaining how the padding is done in the Merkle path and how it is verified in the circuit.

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs
Comment (jpraynaud) [inline mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs:30]: We probably miss some unit tests here to cover edge cases (e.g. Merkle path is effectively padded or not, ...).

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs (hunk 11 above)
Comment (damrobi) [reply mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs]: Done

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs (hunk 12 above)
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs:230]: Removed

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs:220]: As discussed with @jpraynaud we will leave this change for another PR that will investigate in more details the implication of the change.

Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: Similarly to above, those changes go together so it will be done in another PR.

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs (hunk 13 above)
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs:627]: Done

File: mithril-stm/src/proof_system/halo2_snark/mod.rs (hunk 14 above)
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/mod.rs]: Done

File: mithril-stm/src/proof_system/halo2_snark/proof.rs (hunk 16 above)
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/proof.rs:306]: This should just be a parameter so it only impacts the circuit and adds a few constraints. I will check that is does not impact the runtime, otherwise I will reduce it.

File: mithril-stm/src/proof_system/halo2_snark/proof.rs (hunk 15 above)
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/proof.rs:371]: Added

File: mithril-stm/src/proof_system/halo2_snark/prover_input.rs (hunk 17 above)
Comment (damrobi) [reply mithril-stm/src/proof_system/halo2_snark/prover_input.rs:147]: Since we don't have golden test for witness generation I think it didn't break anything (that we cover in the tests). In the end it just adds some zeros to the path so most tests will not be impacted. I can add some golden tests for the witness generation!

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs
Comment (damrobi) [reply mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs:30]: I added some tests in the merkle path gadget file to try to cover the potential behaviors of the modifications.

File: mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs (hunk 18 above)
Comment (damrobi) [reply mithril-stm/src/circuits/halo2/gadgets/merkle_path.rs]: I tried to add a more detailed comment, let me know if it makes it easier to understand or if I should add more details/explanations

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs]: Done

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs [hunk 19]
```diff
@@ -219,12 +219,31 @@ impl<D: Digest + FixedOutput, L: MerkleTreeLeaf> MerkleTree<D, L> {
...
-            proof.push(h.clone());
+            proof.push(h);
             idx = parent(idx);
         }
 
         MerklePath::new(proof, i)
     }
+
+    #[cfg(feature = "future_snark")]
+    /// Get a path (hashes of siblings of the path to the root node)
```
Comment (jpraynaud) [inline mithril-stm/src/membership_commitment/merkle_tree/tree.rs:230]: Could you briefly explain how the padding is done on the Merkle path?

**Review by jpraynaud — APPROVED** [review]
LGTM 👍

File: mithril-stm/src/membership_commitment/merkle_tree/tree.rs (hunk 19 above)
Comment (damrobi) [reply mithril-stm/src/membership_commitment/merkle_tree/tree.rs:230]: I added a more detailed comment

**Review by hjeljeli32 — APPROVED** [review]
LGTM 🚀

**Review by curiecrypt — APPROVED** [review]
LGTM 🚀
